components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,268 @@
+        -> void {
+    archive_writer.write(packed_metadata.str().data(), packed_metadata.str().size());
+}
+
+auto write_archive_file(std::filesystem::path const& file_path, FileWriter& archive_writer)
+        -> void {
+    FileReader reader(file_path.string());
+    std::array<char, cReadBlockSize> read_buffer{};
+    while (true) {
+        size_t num_bytes_read{};
+        ErrorCode const error_code
+                = reader.try_read(read_buffer.data(), cReadBlockSize, num_bytes_read);
+        if (ErrorCode_EndOfFile == error_code) {
+            break;
+        }
+        if (ErrorCode_Success != error_code) {
+            throw OperationFailed(error_code, __FILENAME__, __LINE__);
+        }
+        archive_writer.write(read_buffer.data(), num_bytes_read);
+    }
+}
+
+auto write_archive_files(
+        FileWriter& archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> void {
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        std::filesystem::path static_archive_file_path
+                = multi_file_archive_path / static_archive_file_name;
+        write_archive_file(static_archive_file_path, archive_writer);
+    }
+
+    std::filesystem::path segment_dir_path = multi_file_archive_path / cSegmentsDirname;
+    for (size_t i = 0; i < next_segment_id; ++i) {
+        std::filesystem::path segment_path = segment_dir_path / std::to_string(i);
+        write_archive_file(segment_path, archive_writer);
+    }
+}
+}  // namespace
+
+auto write_single_file_archive(
+        ArchiveMetadata const& multi_file_archive_metadata,
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> void {
+    FileWriter archive_writer;
+    std::filesystem::path single_file_archive_path
+            = multi_file_archive_path.string()
+              + std::string(single_file_archive::cUnstructuredSfaExtension);
+
+    if (std::filesystem::exists(single_file_archive_path)) {
+        throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
+    }
+
+    archive_writer.open(
+            single_file_archive_path.string(),
+            FileWriter::OpenMode::CREATE_FOR_WRITING
+    );
+
+    auto const packed_metadata = pack_single_file_archive_metadata(
+            multi_file_archive_metadata,
+            multi_file_archive_path,
+            next_segment_id
+    );
+
+    write_archive_header(archive_writer, packed_metadata.str().size());
+    write_archive_metadata(archive_writer, packed_metadata);
+    write_archive_files(archive_writer, multi_file_archive_path, next_segment_id);
+
+    archive_writer.close();
+    try {
+        std::filesystem::remove_all(multi_file_archive_path);
+    } catch (std::filesystem::filesystem_error& e) {
+        throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
```

Comment:
Can you try to see if an exception is throw here, what will user see? 

There are cases where the exception is caught and rethrown by some upper level caller, and as a result it's really hard to root cause the issue just by looking at the exception text. Want to avoid this case.

Reply:
@kirkrodrigues wonder if we should strictly throw an exception here, or report an warning.
Technically, the singile file archive is correctly created, we just fail to remove mulit-archive files.

Reply:
Haven't looked at the code, but if the only issue is unnecessary files on disk, sure we could change it to a warning log.